### PR TITLE
Use <a href> instead of <span>, with 302 redirection for liveResolverQuery urls

### PR DESCRIPTION
**Pull request by josephfrazier**
This depends on https://github.com/OctoLinker/live-resolver/pull/12 being merged.

As discussed in https://github.com/OctoLinker/browser-extension/issues/49#issuecomment-137948995.

This is a proof-of-concept and may need to be refined to handle error/edge cases.

### CODE REVIEW

File: lib/insert-link.js [hunk 1]
```diff
@@ -19,6 +20,12 @@ function createLinkElement(text, dataAttr = {}) {
     }
   }
 
+  if (dataAttr.resolver == 'liveResolverQuery') {
+    const type = dataAttr.type;
+    const target = dataAttr.target;
+    linkEl.href = liveResolverQuery({type, target}, true).url;
+  }
```
Comment (stefanbuck) [inline lib/insert-link.js]: In same cases the liveResolver have to be called multiple times for the same target. For example in a JavaScript file like `require('foo')` the target `foo` could be either a bower or npm module.

Maybe the [live-resolver](https://github.com/OctoLinker/live-resolver) should accept multiple types like `https://githublinker.herokuapp.com/q/?type=npm,bower&target=foo`

**Comment by stefanbuck** [comment]
First thanks for picking up this issue again. Ideally every OctoLinker link should be an anchor element so it will behave like every other link on the web. This PR will changes this for some of them and other will stay the same and here is where I see a problem. Imagine the follow scenario. A JavaScript file with two import statements where the first one imports jquery and the second one loads a local file. Our example user prefers to open all links in a new tab with the content menu (right click > Open link in new tab). This would work for the first link, because this is an anchor element. But for the second one, it won't work like the previous one, because it still just a span element. 

I'm not sure if this is a real issue or just me, but UX wise this is not coincident and could confuse the user. What do you think?

File: lib/insert-link.js (hunk 1 above)
Comment (josephfrazier) [inline lib/insert-link.js]: Oh, I hadn't considered that case. Yeah, it sounds like we should use the `tryLoad` logic in the live-resolver as well. I'll probably try to put together a pull request with your proposed syntax sometime this week or over the weekend.

**Comment by josephfrazier** [comment]
That's a good point about UX inconsistency. I had initially threw together this PR as just a demonstration for https://github.com/OctoLinker/live-resolver/pull/12, but I definitely agree that the proposed functionality should be "all or nothing", applying to every link in the same way.

I'm starting to think the best way to do this is to copy _all_ resolution logic into live-resolver, so that we can always rely on having a single `href` that will work regardless of how many possible resolutions there are. Along those lines, maybe it would make sense to combine `live-resolver` into this repo, to make it easier to share logic between the client-side resolver and the server-side resolver? Alternatively, we could do away with the client-side resolution entirely

**Comment by stefanbuck** [comment]
Sorry, busy days for me. Will respond to you beginning of next week.

Cheers Stefan

**Comment by josephfrazier** [comment]
No worries, I've been busy as well. Enjoy the weekend!

**Comment by stefanbuck** [comment]
Hi @josephfrazier,

while developing the latest version of the OctoLinker I had the same idea of moving everything to the `live-resolver`, but there a couple of issues with this approach.

If a user clicks on a file import statement like `import foo from './foo` the extension does a bunch of HEAD request to find the linked file. This is need, because especially for JavaScript the linked resource could be one of the following:

- `./foo.js`
- './foo/index.js`
- ./foo/index.jsx'
- './foo.json`

This is implemented in [resolver/javascript-file](https://github.com/OctoLinker/browser-extension/blob/5bbe5d829cb95b9fe127e0e81d19acf1827ffc6f/lib/resolver/javascript-file.js). 

Also if we would do this on the server, github would properly ban the server, because of too many requests from the same IP.

Another disadvantage of this approach is that it limits this features for public repositories. 

Finally this would generate a lot of traffic on my Heroku and if the server is down for some reasons the extension would be completely useless.


Your second idea of combining the live-resolver into this repo is also not perfect for the following reasons. The extensions would require more permissions in order to make requests to all the different registries. Right from the beginning I was aiming for less as possible permission for this extension to make it more attractive for the user. Further the idea of the current release is to support more and more languages and registries. Every time when a new language is added chrome would ask for new permissions before updating the extensions. Besides that, the user wouldn't benefit from caching already requested packages.

You see, it's not that easy to come up with a solution which address all the mention issues. Nonetheless , I'm grateful for your input!

**Comment by josephfrazier** [comment]
Thanks for the explaining how the local file resolution works. I wasn't aware of that at the time, and it makes sense why that shouldn't be done on the server.

When I suggested combining the live-resolver into this repo, I didn't mean that all of the resolution for external packages should be done client-side (though that is an interesting idea). Instead, what I should have explained is that since the code is closely coupled across repos, it might make sense make sense to combine them into one repo. We could still deploy the extension and server separately, but if there were to be any shared logic, it would be easier to reuse code for it than if the extension code and the server code were in separate repos.

What if we were to add an "eager" mode to the extension? This would be disabled by default, but we could add an [options page](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Anatomy_of_a_WebExtension#Options_pages) for enabling it. If enabled, the extension would attempt to add an href to every link during the extension loading process, whether by resolving local files on the client-side (using the HEAD mechanism you described), or by linking to the server-side resolution (https://github.com/OctoLinker/live-resolver/pull/12) for external packages.

**Comment by stefanbuck** [comment]
Which problem tries this "eager" mode to solve? However, this would put a lot of traffic to the service and even worse the initial load time would increase a lot. For example, a page with ten potential octolinker links and assuming each request takes ~300ms the total load time would be 3 seconds. 

Regarding the mono repo. This is kind of a mono repo already with the `/packages` directory so in general I'm open to extend it. However as you already mentioned, this only make sense if we share server code with the extension. I would like to turn off the server, but I'm not sure how beneficial this will be for us and our users. Why should we do that? Just for a proper link?

**Comment by stefanbuck** [comment]
Closing this because of your comment in https://github.com/OctoLinker/live-resolver/pull/12#issuecomment-263438970

**Pull request closed without merging**